src/content/en/tools/workbox/modules/workbox-sw.md

```diff
@@ -66,7 +66,72 @@ workbox.setConfig({
 });
 ```
 
-With this, you’ll use only the local workbox files.
+With this, you’ll use only the local Workbox files.
+
+## Avoid Async Imports
+
+Under the hood, loading new modules for the first time involves calling
+[`importScripts()`](https://developer.mozilla.org/en-US/docs/Web/API/WorkerGlobalScope/importScripts)
+with the path to the corresponding JavaScript file (either hosted on the CDN, or via a local URL).
+In either case, an important restriction applies: the implicit calls to `importScripts()` can only
+happen inside of a service worker's `install` handler, *or* during the synchronous,
+[initial execution](https://stackoverflow.com/questions/38835273) of the service worker script.
+
+In order to avoid violating this restriction, a best practice is to reference the various
+`workbox.*` namespaces outside of any event handlers or asynchronous functions.
+
+For example, the following top-level service worker code is fine:
+
+```js
+importScripts('{% include "web/tools/workbox/_shared/workbox-sw-cdn-url.html" %}');
+
+// This will work!
+workbox.routing.registerRoute(
+  new RegExp('\.png$'),
+  workbox.strategies.cacheFirst()
+);
+```
+
+But this code could be a problem if you have not referenced `workbox.strategies` elsewhere in your
+service worker:
+
+```js
+importScripts('{% include "web/tools/workbox/_shared/workbox-sw-cdn-url.html" %}');
+
+self.addEventListener('fetch', (event) => {
+  if (event.request.url.endsWith('.png') {
+    // Oops! This causes workbox-strategies.js to be imported inside a fetch handler,
+    // outside of the initial, synchronous service worker execution.
+    const cacheFirst = workbox.strategies.cacheFirst();
+    event.respondWith(cacheFirst.makeRequest({request: event.request}));
+  }
+});
+```
+
+If you need to write code that would otherwise run afoul of this restriction, you can explicitly
+trigger the `importScripts()` call outside of the event handler, using the
+[`workbox.loadModule()`](/web/tools/workbox/reference-docs/latest/workbox#.loadModule) method:
+
+```js
+importScripts('{% include "web/tools/workbox/_shared/workbox-sw-cdn-url.html" %}');
+
+// This will trigger the importScripts() needed for workbox.core and workbox.strategies
+workbox.loadModule('workbox-core');
+workbox.loadModule('workbox-strategies');
```

Comment:
Hmmm, this doesn't seem like the best workaround to me.

The problem with the `loadModule()` method is it doesn't load dependencies, which means the developer has to know what they are, but I don't think we can/should expect them to know that.

I think a much better option is to recommend that they just reference them in the top-level scope, e.g.:

```js
importScripts('https://storage.googleapis.com/workbox-cdn/releases/3.4.1/workbox-sw.js');

const {strategies, routing, googleAnalytics} = workbox;
```

Alternatively we could make `loadModule()` dependency-aware, but I still think it's cleaner to reference the modules in the top-level scope.

Reply:
Here's an example of a SW that just does `workbox.loadModule('workbox-google-analytics');`, and that pulls in a bunch of the related other libraries. I don't think that's any different than what's loaded with your approach.

https://rawgit.com/jeffposnick/4c62a8815b679fd22fb24d57607b6f89/raw/f8c2bb303e50895c33bbd1fe1eb0fa1be9d086c2/index.html

Reply:
(I'm going to remove the call to `workbox.loadModule('workbox-core');` from the example because I realize that's not needed.)

Reply:
Ahhh, yeah the reference to `workbox-core` is what made me think that, and then I looked at the implementation of `loadModule()` and didn't see any dependency references, so I jumped to that conclusion.

But I guess since the build version of `workbox-google-analytics` references `workbox.core`, `workbox.backgroundSync`, etc. it works.

Reply:
What if we document both approaches? That might also help explain with examples exactly what the `workbox-sw` loader is doing.

Reply:
Okay, I put in another example showing what you describe.